README.md
# Web Starter Yeoman Generator [![Build Status](https://travis.ibm.com/arf/generator-web.svg?token=ePBWPJTgR2KYCeTsit1a&branch=master)](https://travis.ibm.com/arf/generator-web) [![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

# 

Creates web project sources in an existing project directory adding boilerplate code for modern web development.

## Getting Started with a generated project

1. Run the generator and choose the options:
  `yo web`

1. Select your chosen framework: None, Angular or React

1. Select your chosen language: Swift, Node, Java, Python-Flask or Python-Django
  
1. Open `public/index.html` in your browser

## Frameworks

Frameworks are chosen using the `framework` option at the end of the generator. For example,

```
yo web --framework {None, React, AngularJS}
```

### Basic

In the basic web project, the following files are created and stored in the `public` directory:

- `public/index.html` - Basic landing page with inline CSS and encoded image assets
- `public/404.html` - Basic 404 error page
- `public/500.html` - Basic 500 error page

It is important that the web server is registered to serve files in the `public` directory.

## React

- [react](https://facebook.github.io/react/) - to build user interfaces
- [webpack](https://webpack.github.io/) - bundles your JS modules
  - sass-loader 
  - css-loader
  - style-loader

Web sources for React projects are stored in the `client` directory:

- `client/index.html` and `client/index.jsx` - Landing page sources
- `client/404.html` and `client/404.jsx` (and corresponding 500 page sources) - Error page sources
- `client/default.css` - minimized stylesheet with inline encoded images

## AngularJS

Web sources for AngularJS projects are stored in the `client` directory:

- `client/index.html` and `client/component.html` - Landing page sources
- `client/404.html` and `client/500.html` - Error pages sources
- `client/default.css` - minimized stylesheet with inline encoded images
- `client/app.js` - Angular app initialization, config and routing


## Publishing Changes

In order to publish changes, you will need to fork the repository or branch off the `master` branch.

Make sure to follow the [conventional commit specification](https://conventionalcommits.org/) before contributing. To help you with commit a commit template is provide. Run `config.sh` to initialize the commit template to your `.git/config` or use [commitizen](https://www.npmjs.com/package/commitizen)

Once you are finished with your changes, run `npm test` to make sure all tests pass.

Do a pull request against `master`, make sure the build passes. A team member will review and merge your pull request.
Once merged to `master` an auto generated pull request will be created against master to update the changelog. Make sure that the CHANGELOG.md and the package.json is correct before merging the pull request. After the auto generated pull request has been merged to `master` the version will be bumped and published to Artifactory.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.3.7"></a>
## [0.3.7](https://github.ibm.com/arf/generator-web/compare/v0.3.6...v0.3.7) (2018-01-18)



<a name="0.3.6"></a>
## [0.3.6](https://github.ibm.com/arf/generator-web/compare/v0.3.5...v0.3.6) (2018-01-12)


### Bug Fixes

* **webpack:** add host attribute 0.0.0.0 to webdev-server ([38bf5ed](https://github.ibm.com/arf/generator-web/commit/38bf5ed))



<a name="0.3.5"></a>
## [0.3.5](https://github.ibm.com/arf/generator-web/compare/v0.3.4...v0.3.5) (2017-12-18)


### Bug Fixes

* fixed a bug with Python flask template ([23ce42b](https://github.ibm.com/arf/generator-web/commit/23ce42b))



<a name="0.3.4"></a>
## [0.3.4](https://github.ibm.com/arf/generator-web/compare/v0.3.3...v0.3.4) (2017-12-08)


### Bug Fixes

* added files for new splash page for webbasic ([a9d1818](https://github.ibm.com/arf/generator-web/commit/a9d1818))



<a name="0.3.3"></a>
## [0.3.3](https://github.ibm.com/arf/generator-web/compare/v0.3.2...v0.3.3) (2017-12-01)


### Bug Fixes

* added files for new splash page for webbasic ([aa2f6e2](https://github.ibm.com/arf/generator-web/commit/aa2f6e2)), closes [arf/planning-node#370](https://github.ibm.com/arf/planning-node/issues/370)



<a name="0.3.2"></a>
## [0.3.2](https://github.ibm.com/arf/generator-web/compare/v0.3.1...v0.3.2) (2017-11-30)


### Bug Fixes

* Dockerfile build script ([bb9f2d3](https://github.ibm.com/arf/generator-web/commit/bb9f2d3))



<a name="0.3.1"></a>
## [0.3.1](https://github.ibm.com/arf/generator-web/compare/v0.3.0...v0.3.1) (2017-11-29)


### Bug Fixes

* add correct client reload script when running npm run dev ([26b8188](https://github.ibm.com/arf/generator-web/commit/26b8188))



<a name="0.3.0"></a>
# [0.3.0](https://github.ibm.com/arf/generator-web/compare/v0.2.3...v0.3.0) (2017-11-29)


### Features

* adding angular framework ([238871c](https://github.ibm.com/arf/generator-web/commit/238871c)), closes [arf/planning-node#57](https://github.ibm.com/arf/planning-node/issues/57)



<a name="0.2.3"></a>
## [0.2.3](https://github.ibm.com/arf/generator-web/compare/v0.2.2...v0.2.3) (2017-11-21)


### Bug Fixes

* react package.json ([9e5aea9](https://github.ibm.com/arf/generator-web/commit/9e5aea9))



<a name="0.2.2"></a>
## [0.2.2](https://github.ibm.com/arf/generator-web/compare/v0.2.1...v0.2.2) (2017-11-21)


### Bug Fixes

* build script ([0eadb87](https://github.ibm.com/arf/generator-web/commit/0eadb87))



<a name="0.2.1"></a>
## [0.2.1](https://github.ibm.com/arf/generator-web/compare/v0.2.0...v0.2.1) (2017-11-21)


### Bug Fixes

* missing procfile ([a232766](https://github.ibm.com/arf/generator-web/commit/a232766))



<a name="0.2.0"></a>
# [0.2.0](https://github.ibm.com/arf/generator-web/compare/v0.1.2...v0.2.0) (2017-11-16)


### Features

* concurrent server and client hot reload for react ([f279ee5](https://github.ibm.com/arf/generator-web/commit/f279ee5))



<a name="0.1.2"></a>
## [0.1.2](https://github.ibm.com/arf/generator-web/compare/v0.1.1...v0.1.2) (2017-11-14)


### Bug Fixes

* updated build process and clarified in README ([c0111fd](https://github.ibm.com/arf/generator-web/commit/c0111fd))



<a name="0.1.1"></a>
## [0.1.1](https://github.ibm.com/arf/generator-web/compare/v0.1.0...v0.1.1) (2017-11-02)

### Bug Fixes

* Rebranding ([9de2f03](https://github.ibm.com/arf/generator-web/commit/9de2f03f3a158c2ad42e8904f82409e2588e71c3))

<a name="0.1.0"></a>
# [0.1.0](https://github.ibm.com/arf/generator-web/compare/v0.0.36...v0.1.0) (2017-11-01)


### Features

* added standard version ([af28e99](https://github.ibm.com/arf/generator-web/commit/af28e99))


test/test.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

"use strict";

let helpers = require('yeoman-test');
let assert = require('yeoman-assert');
let path = require('path');
let fs = require('fs-extra');

let bluemixSettings = {
	name: "MyTest",
	backendPlatform: "SWIFT",
	"cloudant": [
		{
			"password": "pass",
			"url": "https://account.cloudant.com",
			"username": "user"
		}
	],
	"objectStorage": [
		{
			"password": "Gl.=W23@",
			"projectId": "12345",
			"region": "dallas",
			"userId": "abc1234"
		}
	]
};

let requiredFilesForBasic = [
	'public/index.html',
	'public/js/bundle.js',
	'public/css/default.css'
];

let requiredFilesForAngular = [
	'client/app.js',
	'client/component.html',
	'client/index.html',
	'test/test-server.js',
	'webpack.common.js',
	'webpack.dev-proxy.js',
	'webpack.dev-standalone.js',
	'webpack.prod.js',
];

let requiredFilesForWebpack = [
	'public/index.html',
	'src/client/app/index.jsx',
	'src/client/sass/default.scss',
	'webpack.config.js'
];

let requiredFilesForReact = [
	'client/index.html',
	'client/app/App.jsx',
	'client/default.scss',
	'test/test-server.js',
	'webpack.common.js',
	'webpack.dev-proxy.js',
	'webpack.dev-standalone.js',
	'webpack.prod.js',
];

describe('Web project generator', function () {

	// NodeJS tests

	describe('Basic app with NodeJS', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir()
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "None"
				});
		});

		it('does not contain web pages', function () {

			assert.noFile(requiredFilesForBasic);

		});

		it('does not contains web sources', function () {
			assert.noFile(['src/client/index.html',
				'src/client/sass/default.scss'
			]);
		});

		// it('starter text appears', function () {
		// 	assert.fileContent('public/index.html', 'You are currently running a NodeJS server.')
		// });


	});

	describe('Webpack app with NodeJS', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/package.json'), path.join(dir, 'package.json'));
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "Webpack"
				});
		});

		it('required files created', function () {

			assert.file(requiredFilesForWebpack);

		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});

		it('contains new dependencies', function () {

			assert.fileContent('package.json', 'webpack');
			assert.fileContent('package.json', 'babel');

		});

		it('does not have React', function () {
			assert.noFileContent('package.json', 'react');
			assert.noFileContent('package.json', 'react-dom');
		});

		it('starter writes NodeJS', function () {
			assert.fileContent('src/client/app/index.jsx', 'You are currently running a NodeJS server.');
		});

		it('should modify Dockerfile', function () {
			assert.fileContent('Dockerfile', 'npm run build;');
		});

		it('should modify manifest.yml', function () {
			assert.fileContent('manifest.yml', 'npm prune --production');
			assert.fileContent('manifest.yml', 'NPM_CONFIG_PRODUCTION');
		})
	});

	describe('React app with NodeJS', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/package.json'), path.join(dir, 'package.json'));
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "React"
				});
		});

		it('required files created', function () {

			assert.file(requiredFilesForReact);

		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});

		it('contains Webpack', function () {

			assert.fileContent('package.json', 'webpack');
			assert.fileContent('package.json', 'babel');

			// assert.fileContent('package.json', 'webpack-dev-server');

		});

		it('contains React', function () {
			assert.fileContent('package.json', 'react');
			assert.fileContent('package.json', 'react-dom');
			assert.fileContent('package.json', 'babel-preset-react');
		});

		it('should modify Dockerfile', function () {
			assert.fileContent('Dockerfile', 'npm run build;');
		});

		it('should modify manifest.yml', function () {
			assert.fileContent('manifest.yml', 'npm prune --production');
			assert.fileContent('manifest.yml', 'NPM_CONFIG_PRODUCTION');
		})
	});

	describe('AngularJS app with NodeJS and present existing files', function () {
		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/package.json'), path.join(dir, 'package.json'));
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "AngularJS"
				});
		});

		it('required files created', function () {
			assert.file(requiredFilesForAngular);
		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});


		it('should modify Dockerfile', function () {
			assert.fileContent('Dockerfile', 'npm run build;');
		});

		it('should modify manifest.yml', function () {
			assert.fileContent('manifest.yml', 'npm prune --production');
			assert.fileContent('manifest.yml', 'NPM_CONFIG_PRODUCTION');
		})

		it('should have react specific build script', function() {
			assert.fileContent('package.json', 'webpack --progress --config webpack.prod.js');
		})

		it('should have original scripts and dependencies', function() {
			assert.fileContent('package.json', 'mocha');
			assert.fileContent('package.json', 'node --debug server/server.js');
		})
	});

	describe('AngularJS app with NodeJS with empty user directory', function () {
		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "AngularJS"
				});
		});

		it('required files created', function () {
			assert.file(requiredFilesForAngular);
		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});


		it('should not have Dockerfile', function () {
			assert.noFile('Dockerfile')
		});

		it('should not have manifest.yml', function () {
			assert.noFile('manifest.yml');
		});

		it('should have react specific build script', function() {
			assert.fileContent('package.json', 'webpack --progress --config webpack.prod.js');
		});

		it('should have original scripts and dependencies', function() {
			assert.fileContent('package.json', 'mocha');
			assert.fileContent('package.json', 'node --debug server/server.js');
		})
	});
});


generators/app/templates/basic/test-server.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

var expect = require('chai').expect;
var http = require('http');

// Below code demonstrates using various methods of testing
describe('Testing Server', function() {

  before(function(done){
    require(process.cwd() + '/server/server');
    setTimeout(done, 5000); // Waiting 5 seconds for server to start
    this.timeout(10000);
  });
    it('Public endpoint returns "Hello!"', function(done){
      var responseString = '';

      var options = {
        host: 'localhost',
        port: process.env.PORT || 3000,
        path: '/'
      };

      var callback = function(response){
        response.on('data', function (chunk) {
          responseString += chunk;
        });

        response.on('end', function () {
          expect(responseString).to.include('You are currently running a NodeJS server.');
          done();
        });
      };

      http.request(options, callback).end();
    });

    it('Health endpoint shows status up', function(done){
      var responseString = '';

      var options = {
        host: 'localhost',
        port: process.env.PORT || 3000,
        path: '/health'
      };

      var callback = function(response){
        response.on('data', function (chunk) {
          responseString += chunk;
        });

        response.on('end', function () {
          expect(responseString).to.equal('{"status":"UP"}');
          done();
        });
      };

      http.request(options, callback).end();
    });
});


generators/app/templates/react/webpack.common.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const CopyWebpackPlugin = require('copy-webpack-plugin');

const WebpackPluginPluginConfig = new HtmlWebpackPlugin({
  template: './client/index.html',
  filename: 'index.html',
  inject: 'body'
});

const CopyWebpackPluginConfig= new CopyWebpackPlugin([
  { from: './client/default.css', to: path.resolve(__dirname, 'public') },
  { from: './client/404.html', to: path.resolve(__dirname, 'public') },
  { from: './client/500.html', to: path.resolve(__dirname, 'public') }
]);

module.exports = {
  entry: './client/index.jsx',
  output: {
    filename: 'js/bundle.[hash].js',
    path: path.resolve(__dirname, 'public')
  },
  module : {
    loaders : [
      {
          test    : /\.jsx?$/,
          exclude : /node_modules/,
          loader  : 'babel-loader'
      },
      {
          test    : /\.css?/,
          loader  : 'style-loader!css-loader!sass-loader'
      }
    ]
  },
  plugins: [
    WebpackPluginPluginConfig,
    CopyWebpackPluginConfig
  ]
};

generators/app/templates/angularjs/webpack.prod.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

const webpack = require('webpack');
const merge = require('webpack-merge');
const common = require('./webpack.common');

module.exports = merge(common, {
  devtool: 'source-map',
  plugins: [
    new webpack.optimize.UglifyJsPlugin(),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': '"production"'
    })
  ]
});

generators/app/templates/angularjs/webpack.dev-proxy.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

const merge = require('webpack-merge');
const common = require('./webpack.common');

module.exports = merge(common, {
  devtool: 'inline-source-map',
  devServer: {
    port: 3000,
    proxy: {
      "/appmetrics-dash": "http://localhost:3100",
      "/health": "http://localhost:3100"
    }
  }
});

generators/app/index.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

 'use strict';
const Generator = require('yeoman-generator');
const yamlParser = require('yaml-parser');
const dep = require('./templates/dependencies.json');
const reactDep = require('./templates/react/dependencies.json');
const angularJsDep = require('./templates/angularjs/dependencies.json');

const scripts = dep.scripts;
const angularJsScripts = angularJsDep.scripts;

const devDependencies = dep.devDependencies;
const angularJsDevDependencies = angularJsDep.devDependencies;

module.exports = class extends Generator {
	constructor(args, opts) {
		super(args, opts);

		this.framework = opts.framework;

		if (typeof (opts.bluemix) === 'string') {
			this.bluemix = JSON.parse(opts.bluemix || "{}");
		} else {
			this.bluemix = opts.bluemix;
		}

		this.humanNameLanguage = {
			"NODE": "NodeJS",
			"SWIFT": "Swift",
			"JAVA": "Java",
			"PYTHON": "Python",
			"DJANGO": "Django"
		};
	}

	paths() {
		this.sourceRoot();
		this.templatePath('index.js');
	}

	prompting() {
		let prompts = [];
		if (this.bluemix === undefined) {
			this.bluemix = {};
			prompts.push({
				type: 'input',
				name: 'name',
				message: 'Your project name'
			});
			prompts.push({
				type: 'list',
				name: 'framework',
				message: 'What framework are you using',
				choices: [
					"None",
					"React",
					"AngularJS"
				]
			});
			prompts.push({
				type: 'list',
				name: 'language',
				message: 'What language are you using',
				choices: [
					"SWIFT",
					"NODE",
					"JAVA",
					"PYTHON",
					"DJANGO"
				]
			});
		}

		return this.prompt(prompts).then(this._processAnswers.bind(this));
	}

	_processAnswers(answers) {

		this.bluemix.backendPlatform = answers.language || this.bluemix.backendPlatform;
		this.framework = answers.framework || this.framework;
		this.bluemix.name = answers.name || this.bluemix.name;
	}

	write() {
		switch (this.framework) {
			case "None":
				this._generateBasic();
				break;
			case "React":
				this._generateReact();
				break;
			case "AngularJS":
				this._generateAngularJS();
				break;
			default:
				throw 'Unknown generator type';
		}
	}

	_generateBasic() {

		// Replace server test with Web specific test.
		if (this.bluemix.backendPlatform === 'NODE') {
			this.fs.copyTpl(
				this.templatePath('basic/test-server.js'),
				this.destinationPath('test/test-server.js'), {}
			);

			this.fs.copyTpl(
				this.templatePath('basic/node'),
				this.destinationPath('public'), {}
			);

		}

		else {
			this.fs.copyTpl(
				this.templatePath('basic/python/index.html'),
				this.destinationPath('public/index.html'), {
					applicationName: this.bluemix.name,
					language: this.humanNameLanguage[this.bluemix.backendPlatform]
				}
			);
			this.fs.copyTpl(
				this.templatePath('basic/python/404.html'),
				this.destinationPath('public/404.html'), {}
			);
			this.fs.copyTpl(
				this.templatePath('basic/python/500.html'),
				this.destinationPath('public/500.html'), {}
			);
		}

	}

	_generateReact() {

		// Replace server test with Web specific test.
		this.fs.copyTpl(
			this.templatePath('react/test-server.js'),
			this.destinationPath('test/test-server.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/Procfile-dev'),
			this.destinationPath('Procfile-dev'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/babelrc'),
			this.destinationPath('.babelrc'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.common.js'),
			this.destinationPath('webpack.common.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.dev-proxy.js'),
			this.destinationPath('webpack.dev-proxy.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.dev-standalone.js'),
			this.destinationPath('webpack.dev-standalone.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.prod.js'),
			this.destinationPath('webpack.prod.js'), {}
		);

		if (!this.fs.exists(this.destinationPath('package.json'))) {
			this.fs.copyTpl(
				this.templatePath('package.json'),
				this.destinationPath('package.json'), {
					applicationName: this.bluemix.name,
					language: this.bluemix.backendPlatform
				}
			);
		}

		this._augmentPackageJSON({ react: true });

		this.fs.copyTpl(
			this.templatePath('react/client'),
			this.destinationPath('client'), {}
		);

		this._rewriteBuild();
	}

	_generateAngularJS() {

		// Replace server test with Web specific test.
		this.fs.copyTpl(
			this.templatePath('angularjs/test-server.js'),
			this.destinationPath('test/test-server.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.common.js'),
			this.destinationPath('webpack.common.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.dev-proxy.js'),
			this.destinationPath('webpack.dev-proxy.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.dev-standalone.js'),
			this.destinationPath('webpack.dev-standalone.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.prod.js'),
			this.destinationPath('webpack.prod.js'), {}
		);

		if (!this.fs.exists(this.destinationPath('package.json'))) {
			this.fs.copyTpl(
				this.templatePath('package.json'),
				this.destinationPath('package.json'), {
					applicationName: this.bluemix.name,
					language: this.bluemix.backendPlatform
				}
			);
		}

		this._augmentPackageJSON({ angularjs: true });

		this.fs.copyTpl(
			this.templatePath('angularjs/client/app.js'),
			this.destinationPath('client/app.js'), {}
		);
		this.fs.copyTpl(
			this.templatePath('angularjs/client/component.html'),
			this.destinationPath('client/component.html'), {}
		);
		this.fs.copyTpl(
			this.templatePath('angularjs/client/default.css'),
			this.destinationPath('client/default.css'), {}
		);
		this.fs.copyTpl(
			this.templatePath('angularjs/client/index.html'),
			this.destinationPath('client/index.html'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/client/404.html'),
			this.destinationPath('public/404.html'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/client/500.html'),
			this.destinationPath('public/500.html'), {}
		);

		this._rewriteBuild();
	}

	_augmentPackageJSON(options) {
		let packageFile = this.fs.read(this.destinationPath('package.json'), { defaults: "{}" });
		let packageFileJSON;

		try {
			packageFileJSON = JSON.parse(packageFile);
		} catch (err) {
			packageFileJSON = {};
		}

		if (packageFileJSON.devDependencies === undefined) {
			packageFileJSON.devDependencies = {};
		}

		packageFileJSON.scripts = Object.assign(packageFileJSON.scripts, scripts);
		packageFileJSON.devDependencies = Object.assign(packageFileJSON.devDependencies, devDependencies);

		if (options !== undefined && options.react) {
			packageFileJSON.scripts = Object.assign(packageFileJSON.scripts, reactDep.scripts);
			packageFileJSON.devDependencies = Object.assign(packageFileJSON.devDependencies, reactDep.devDependencies);
		}

		if (options !== undefined && options.angularjs) {
			packageFileJSON.devDependencies = Object.assign(packageFileJSON.devDependencies, angularJsDevDependencies);
			packageFileJSON.scripts = Object.assign(packageFileJSON.scripts, angularJsScripts);
		}

		this.fs.writeJSON(this.destinationPath('package.json'), packageFileJSON, null, 4);
	}

	_rewriteBuild() {

		if (this.fs.exists(this.destinationPath('Dockerfile'))) {
			let Dockerfile = this.fs.read(this.destinationPath('Dockerfile'));
			if (Dockerfile.indexOf('webpack.config.js') < 0) {
				Dockerfile = Dockerfile.replace('COPY package.json /app/\nRUN cd /app; npm install --production', 'COPY package.json webpack.config.js /app/\nRUN cd /app; npm install --production\nCOPY /client /app/client/\nRUN npm install --only=dev; npm run build; npm prune --production');
			}

			this.fs.write(this.destinationPath('Dockerfile'), Dockerfile);
		}

		if (this.fs.exists(this.destinationPath('manifest.yml'))) {
			let manifest = this.fs.read(this.destinationPath('manifest.yml'));
			if (manifest.indexOf('npm prune') < 0) {
				manifest = manifest.replace('npm start', 'npm prune --production && NODE_ENV=production npm start');
			}

			if (manifest.indexOf('NPM_CONFIG_PRODUCTION' < 0)) {
				let manifest_parsed = yamlParser.safeLoad(manifest);

				if (manifest_parsed.applications && manifest_parsed.applications[0]) {
					let env = manifest_parsed.applications[0].env || {};
					env.NPM_CONFIG_PRODUCTION = false;
					manifest_parsed.applications[0].env = env;
					manifest = yamlParser.safeDump(manifest_parsed);
				}
			}

			this.fs.write(this.destinationPath('manifest.yml'), manifest);
		}
	}

};